File: pubnub-java/src/main/java/com/pubnub/api/models/consumer/pubsub/message_actions/PNMessageActionResult.java
package com.pubnub.api.models.consumer.pubsub.message_actions;

import com.pubnub.api.models.consumer.message_actions.PNMessageAction;
import com.pubnub.api.models.consumer.pubsub.BasePubSubResult;
import com.pubnub.api.models.consumer.pubsub.objects.ObjectResult;
import lombok.Builder;
import lombok.ToString;

@ToString(callSuper = true)
public class PNMessageActionResult extends ObjectResult<PNMessageAction> {

    @Builder(builderMethodName = "actionBuilder")
    private PNMessageActionResult(BasePubSubResult result, String event, PNMessageAction data) {
        super(result, event, data);
    }

    public PNMessageAction getMessageAction() {
        return data;
    }
}

File: src/main/kotlin/com/pubnub/api/models/consumer/pubsub/message_actions/PNMessageActionResult.kt
package com.pubnub.api.models.consumer.pubsub.message_actions

import com.pubnub.api.callbacks.SubscribeCallback
import com.pubnub.api.models.consumer.message_actions.PNMessageAction
import com.pubnub.api.models.consumer.pubsub.BasePubSubResult
import com.pubnub.api.models.consumer.pubsub.PubSubResult
import com.pubnub.api.models.consumer.pubsub.objects.ObjectResult

/**
 * Wrapper around message actions received in [SubscribeCallback.messageAction].
 *
 * @property event The message action event. Could be `added` or `removed`.
 * @property data The actual message action.
 */
data class PNMessageActionResult(
    private val result: BasePubSubResult,
    override val event: String,
    override val data: PNMessageAction
) : ObjectResult<PNMessageAction>, PubSubResult by result {
    val messageAction: PNMessageAction = data
}
